### PR TITLE
AV-185518: AKO does not create VirtualService if hostrule is specified without mentioning listeners.

### PR DESCRIPTION
**Pull request by saurabhvaidvmware**

### CODE REVIEW

**Comment by saurabhvaidvmware** [comment]
build ako

**Comment by saurabhvaidvmware** [comment]
build ako

**Comment by saurabhvaidvmware** [comment]
build ako

File: tests/evhtests/l7_evh_crd_test.go [hunk 1]
```diff
@@ -393,18 +393,21 @@ func TestCreateDeleteSharedVSHostRuleForEvh(t *testing.T) {
 	g.Expect(nodes[0].VsDatascriptRefs).To(gomega.HaveLen(2))
 	g.Expect(nodes[0].VsDatascriptRefs[0]).To(gomega.ContainSubstring("thisisaviref-ds2"))
 	g.Expect(nodes[0].VsDatascriptRefs[1]).To(gomega.ContainSubstring("thisisaviref-ds1"))
-	g.Expect(nodes[0].PortProto).To(gomega.HaveLen(3))
+	g.Expect(nodes[0].PortProto).To(gomega.HaveLen(5))
 	var ports []int
+	sslPorts := [3]int{443, 8083}
```
Comment (akshayhavile) [inline tests/evhtests/l7_evh_crd_test.go:398]: it should be `[2]int{443, 8083}`

File: tests/ingresstests/crd_test.go [hunk 2]
```diff
@@ -157,18 +157,21 @@ func TestCreateDeleteSharedVSHostRule(t *testing.T) {
 	g.Expect(nodes[0].VsDatascriptRefs).To(gomega.HaveLen(2))
 	g.Expect(nodes[0].VsDatascriptRefs[0]).To(gomega.ContainSubstring("thisisaviref-ds2"))
 	g.Expect(nodes[0].VsDatascriptRefs[1]).To(gomega.ContainSubstring("thisisaviref-ds1"))
-	g.Expect(nodes[0].PortProto).To(gomega.HaveLen(3))
+	g.Expect(nodes[0].PortProto).To(gomega.HaveLen(5))
 	var ports []int
+	sslPorts := [3]int{443, 8083}
```
Comment (akshayhavile) [inline tests/ingresstests/crd_test.go:162]: same as above

**Review by akshayhavile — COMMENTED** [review]
Overall changes looks good to me

File: tests/evhtests/l7_evh_crd_test.go (hunk 1 above)
Comment (saurabhvaidvmware) [reply tests/evhtests/l7_evh_crd_test.go]: Changed

**Comment by saurabhvaidvmware** [comment]
build ako

**Comment by saurabhvaidvmware** [comment]
build ako

**Review by aaha97 — APPROVED** [review]
LGTM

**Comment by saurabhvaidvmware** [comment]
build ako

**Comment by saurabhvaidvmware** [comment]
build ako

**Comment by saurabhvaidvmware** [comment]
build ako

File: internal/nodes/crd_translator.go [hunk 3]
```diff
@@ -75,7 +75,10 @@ func BuildL7HostRule(host, key string, vsNode AviVsEvhSniModel) {
 
 	portProtocols := []AviPortHostProtocol{
 		{Port: 80, Protocol: utils.HTTP},
-		{Port: 443, Protocol: utils.HTTP, EnableSSL: true},
+	}
+
+	if vsNode.IsSecure() || !vsNode.IsDedicatedVS() {
+		portProtocols = append(portProtocols, AviPortHostProtocol{Port: 443, Protocol: utils.HTTP, EnableSSL: true})
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply internal/nodes/crd_translator.go:81]: For L4 case, we are setting application profile based on protocol `sctp`, `tcp` and `udp`. But for L7, currently we do not set that. So it has been set has `HTTP` and depending upon `enableSSL` flag we set corresponding field in VS at Controller side.

**Comment by saurabhvaidvmware** [comment]
build ako

**Comment by saurabhvaidvmware** [comment]
build ako

**Review by akshayhavile — APPROVED** [review]

**Review by pkoshtavmware — APPROVED** [review]
LGTM

**Pull request merged**